_posts/2025-06-04-#30605.md

```diff
@@ -10,17 +10,39 @@ status: upcoming
+
+- At a high level, cluster linearization is the process of sorting the transactions within a given cluster in a way that respects dependencies (parents before children) while maximizing the effective fee rate when considering "chunks" of this ordered list. The goal is to produce an optimal linearization, an ordering with the best possible fee rate profile, for the transactions in that cluster. (for more in-depth reading, see [here](https://delvingbitcoin.org/t/how-to-linearize-your-cluster/303))
+    - The data structures and algorithms for cluster linearization can be found in [`cluster_linearize.h`](https://github.com/bitcoin/bitcoin/blob/master/src/cluster_linearize.h). **
+
+- The linearization code is complex and therefore relies on fuzz testing to ensure correctness. Using the wide variety of data generated by the fuzzer to build dependency graphs (`DepGraph` instances) and produce/update linearizations is probably our best bet for finding any subtle bugs or edge cases in the code.
+
+- The `cluster_linearize` fuzz test includes a few helper classes and functions that are used to compare against the actual implementation. The motivation for this PR is to construct new fuzz targets that are used to make sure these helpers work properly. By having dedicated targets that test the test, we can be even more confident that our cluster linearization code is bug-free. Woo!
+
+** There's a [PR](https://github.com/bitcoin/bitcoin/pull/32545) that aims to replace the current algorithm with a new and improved one called spanning-forest linearization. ([in-depth reading](https://delvingbitcoin.org/t/spanning-forest-cluster-linearization/1419)) 
+
+
+### Exercise (Optional)
+
+- Try fuzzing the `clusterlin_linearize` target. You can refer to [fuzzing docs](https://github.com/bitcoin/bitcoin/blob/master/doc/fuzzing.md) for assistance. Using `--preset=libfuzzer-nosan` will likely make it easier.
+
+- Now change `chunking` to `simple_chunking` at this [line](https://github.com/bitcoin/bitcoin/blob/d2fc10b64505512a70bed05e7bb21d76c8f748cd/src/test/fuzz/cluster_linearize.cpp#L1104) and run the target again.
+
+- After a few minutes, you should see a crash. Welcome to fuzzing. This is a simple example of a tiny bug in the test itself which fuzz testing catches very quickly. A lot of the time, a crash happens because something is wrong in the fuzz harness itself, as opposed to a bug in the production code. This is why it can be useful to have targets or assertions that ensure that the fuzz test itself is correct.
+
 
 
 ## Questions
 
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+
+2. There are 14 targets in this fuzz test that build on each other. What is each target fuzzing? Which ones are testing the actual linearization code and which ones are testing the test?
```

Comment:
> What is each target fuzzing?

Letting all attendees describe 14 fuzz targets might be a bit much. Perhaps this might work better as some kind of puzzle, like "Which target is doing X"?